File: src/sorting.py
"""
===================================================================
  IMPLEMENTASI ALGORITMA SORTING - PYTHON
  Algoritma: 1. Bubble Sort, 2. Selection Sort, 3. Insertion Sort,
             4. Quick Sort, 5. Merge Sort
===================================================================
"""

import copy

DIVIDER = "=" * 60
THIN    = "-" * 60


def print_array(arr, highlight=None, label=""):
    """Cetak array dengan highlight indeks tertentu."""
    result = []
    for i, val in enumerate(arr):
        if highlight and i in highlight:
            result.append(f"[{val}]")
        else:
            result.append(f" {val} ")
    line = "  ".join(result)
    if label:
        print(f"  {label:<25} {line}")
    else:
        print(f"  {line}")


#===================
# 1. BUBBLE SORT ||
#===================

def bubble_sort(arr):
    print(f"\n{DIVIDER}")
    print("  BUBBLE SORT")
    print(DIVIDER)
    a = arr[:]
    n = len(a)
    iteration = 0

    print(f"  Data Awal  : {a}")
    print(THIN)

    for i in range(n - 1):
        swapped = False
        for j in range(n - 1 - i):
            iteration += 1
            if a[j] > a[j + 1]:
                a[j], a[j + 1] = a[j + 1], a[j]
                swapped = True
            print(f"  Pass {i+1}, Step {j+1} (iterasi {iteration:>3}): {a}  "
                  f"← bandingkan indeks {j} & {j+1}")
        if not swapped:
            print(f"  [Tidak ada pertukaran pada pass {i+1}, berhenti lebih awal]")
            break
        print(f"  Setelah Pass {i+1}     : {a}")
        print()

    print(THIN)
    print(f"  Total iterasi : {iteration}")
    print(f"  Hasil Sorting : {a}")
    return a

#=====================
# 2. SELECTION SORT ||
#=====================

def selection_sort(arr):
    print(f"\n{DIVIDER}")
    print("  SELECTION SORT")
    print(DIVIDER)
    a = arr[:]
    n = len(a)
    iteration = 0

    print(f"  Data Awal  : {a}")
    print(THIN)

    for i in range(n - 1):
        min_idx = i
        for j in range(i + 1, n):
            iteration += 1
            if a[j] < a[min_idx]:
                min_idx = j
            print(f"  Pass {i+1}, Step {j-i:>2} (iterasi {iteration:>3}): "
                  f"bandingkan indeks {i}({a[i]}) dengan {j}({a[j]})  "
                  f"→ min sementara = a[{min_idx}]={a[min_idx]}")

        if min_idx != i:
            a[i], a[min_idx] = a[min_idx], a[i]
            print(f"  Tukar a[{i}]={a[i]} ↔ a[{min_idx}]={a[min_idx]}")
        else:
            print(f"  Tidak ada pertukaran (a[{i}]={a[i]} sudah minimum)")
        print(f"  Setelah Pass {i+1}     : {a}")
        print()

    print(THIN)
    print(f"  Total iterasi : {iteration}")
    print(f"  Hasil Sorting : {a}")
    return a

#=====================
# 3. INSERTION SORT ||
#=====================

def insertion_sort(arr):
    print(f"\n{DIVIDER}")
    print("  INSERTION SORT")
    print(DIVIDER)
    a = arr[:]
    n = len(a)
    iteration = 0

    print(f"  Data Awal  : {a}")
    print(THIN)

    for i in range(1, n):
        key = a[i]
        j = i - 1
        print(f"  Pass {i}: ambil kunci = {key}")
        while j >= 0 and a[j] > key:
            iteration += 1
            a[j + 1] = a[j]
            print(f"    iterasi {iteration:>3}: geser a[{j}]={a[j+1]} ke kanan → {a}")
            j -= 1
        iteration += 1
        a[j + 1] = key
        print(f"    iterasi {iteration:>3}: sisipkan {key} di posisi {j+1}  → {a}")
        print()

    print(THIN)
    print(f"  Total iterasi : {iteration}")
    print(f"  Hasil Sorting : {a}")
    return a

#=================
# 4. QUICK SORT ||
#=================

_qs_iter = [0]

def _partition(a, low, high, depth):
    pivot = a[high]
    i = low - 1
    print(f"  {'  ' * depth}Pivot = {pivot}  subarray = {a[low:high+1]}")
    for j in range(low, high):
        _qs_iter[0] += 1
        if a[j] <= pivot:
            i += 1
            a[i], a[j] = a[j], a[i]
            print(f"  {'  ' * depth}  iterasi {_qs_iter[0]:>3}: a[{i}]↔a[{j}]  → {a[low:high+1]}")
        else:
            print(f"  {'  ' * depth}  iterasi {_qs_iter[0]:>3}: lewati a[{j}]={a[j]} (> pivot {pivot})")
        _qs_iter[0] += 1
    a[i + 1], a[high] = a[high], a[i + 1]
    print(f"  {'  ' * depth}  Letakkan pivot {pivot} di posisi {i+1}  → {a[low:high+1]}")
    return i + 1

def _quick_sort(a, low, high, depth=0):
    if low < high:
        pi = _partition(a, low, high, depth)
        print(f"  {'  ' * depth}Rekursif kiri  [{low}..{pi-1}]: {a[low:pi]}")
        _quick_sort(a, low, pi - 1, depth + 1)
        print(f"  {'  ' * depth}Rekursif kanan [{pi+1}..{high}]: {a[pi+1:high+1]}")
        _quick_sort(a, pi + 1, high, depth + 1)

def quick_sort(arr):
    print(f"\n{DIVIDER}")
    print("  QUICK SORT")
    print(DIVIDER)
    a = arr[:]
    _qs_iter[0] = 0

    print(f"  Data Awal  : {a}")
    print(THIN)

    _quick_sort(a, 0, len(a) - 1)

    print(THIN)
    print(f"  Total iterasi : {_qs_iter[0]}")
    print(f"  Hasil Sorting : {a}")
    return a

#=================
# 5. MERGE SORT ||
#=================

_ms_iter = [0]

def _merge(a, left, mid, right, depth):
    L = a[left:mid + 1]
    R = a[mid + 1:right + 1]
    i = j = 0
    k = left
    print(f"  {'  ' * depth}Gabung: {L} + {R}")
    while i < len(L) and j < len(R):
        _ms_iter[0] += 1
        if L[i] <= R[j]:
            a[k] = L[i]; i += 1
        else:
            a[k] = R[j]; j += 1
        k += 1
    while i < len(L):
        a[k] = L[i]; i += 1; k += 1
    while j < len(R):
        a[k] = R[j]; j += 1; k += 1
    print(f"  {'  ' * depth}Hasil : {a[left:right+1]}")

def _merge_sort(a, left, right, depth=0):
    if left < right:
        mid = (left + right) // 2
        print(f"  {'  ' * depth}Bagi [{left}..{right}] → [{left}..{mid}] | [{mid+1}..{right}]")
        _ms_sort_left  = a[left:mid + 1]
        _ms_sort_right = a[mid + 1:right + 1]
        _merge_sort(a, left, mid, depth + 1)
        _merge_sort(a, mid + 1, right, depth + 1)
        _merge(a, left, mid, right, depth)

def merge_sort(arr):
    print(f"\n{DIVIDER}")
    print("  MERGE SORT")
    print(DIVIDER)
    a = arr[:]
    _ms_iter[0] = 0

    print(f"  Data Awal  : {a}")
    print(THIN)

    _merge_sort(a, 0, len(a) - 1)

    print(THIN)
    print(f"  Total iterasi : {_ms_iter[0]}")
    print(f"  Hasil Sorting : {a}")
    return a

#========
# MAIN ||
#========

def main():
    print(DIVIDER)
    print("  PROGRAM SORTING ALGORITHM - PYTHON")
    print(DIVIDER)

    n = int(input("\n  Masukkan jumlah data : "))
    print(f"  Masukkan {n} data (pisahkan dengan spasi) : ", end="")
    data = list(map(int, input().split()))

    if len(data) != n:
        print(f"  [PERINGATAN] Anda memasukkan {len(data)} data, bukan {n}.")
        data = data[:n]

    print(f"\n  Data yang akan di-sort : {data}")

    print(f"\n{DIVIDER}")
    print("  Pilih Algoritma Sorting:")
    print("  1. Bubble Sort")
    print("  2. Selection Sort")
    print("  3. Insertion Sort")
    print("  4. Quick Sort")
    print("  5. Merge Sort")
    print("  6. Semua Algoritma")
    print(DIVIDER)

    pilihan = input("  Masukkan pilihan (1-6) : ").strip()

    algo_map = {
        "1": bubble_sort,
        "2": selection_sort,
        "3": insertion_sort,
        "4": quick_sort,
        "5": merge_sort,
    }

    if pilihan == "6":
        for fn in algo_map.values():
            fn(data)
    elif pilihan in algo_map:
        algo_map[pilihan](data)
    else:
        print("  Pilihan tidak valid.")

    print(f"\n{DIVIDER}")
    print("  Program selesai.")
    print(DIVIDER)


if __name__ == "__main__":
    main()

File: src/sorting.cpp
/*
===================================================================
  IMPLEMENTASI ALGORITMA SORTING - C++
  Algoritma: 1. Bubble Sort, 2. Selection Sort, 3. Insertion Sort,
             4. Quick Sort, 5. Merge Sort
===================================================================
*/

#include <iostream>
#include <vector>
#include <string>
#include <iomanip>

using namespace std;

 /* ─── UTILITY ─── */

const string DIVIDER(60, '=');
const string THIN(60, '-');

void printArray(const vector<int>& arr, const string& label = "") {
    if (!label.empty())
        cout << "  " << left << setw(25) << label;
    else
        cout << "  ";
    cout << "[ ";
    for (size_t i = 0; i < arr.size(); i++) {
        cout << arr[i];
        if (i < arr.size() - 1) cout << ", ";
    }
    cout << " ]" << endl;
}

void printArrayRange(const vector<int>& arr, int lo, int hi) {
    cout << "[ ";
    for (int i = lo; i <= hi; i++) {
        cout << arr[i];
        if (i < hi) cout << ", ";
    }
    cout << " ]";
}

 /* ─── 1. BUBBLE SORT ─── */

vector<int> bubbleSort(vector<int> arr) {
    cout << "\n" << DIVIDER << "\n";
    cout << "  BUBBLE SORT\n";
    cout << DIVIDER << "\n";

    int n = arr.size();
    int iteration = 0;

    printArray(arr, "Data Awal :");
    cout << THIN << "\n";

    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;
        for (int j = 0; j < n - 1 - i; j++) {
            iteration++;
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
            cout << "  Pass " << i+1 << ", Step " << j+1
                 << " (iterasi " << setw(3) << iteration << "): ";
            printArrayRange(arr, 0, n-1);
            cout << "  <- bandingkan indeks " << j << " & " << j+1 << "\n";
        }
        if (!swapped) {
            cout << "  [Tidak ada pertukaran pada pass " << i+1
                 << ", berhenti lebih awal]\n";
            break;
        }
        cout << "  Setelah Pass " << i+1 << "    : ";
        printArrayRange(arr, 0, n-1);
        cout << "\n\n";
    }

    cout << THIN << "\n";
    cout << "  Total iterasi : " << iteration << "\n";
    printArray(arr, "Hasil Sorting :");
    return arr;
}

 /* ─── 2. SELECTION SORT ─── */

vector<int> selectionSort(vector<int> arr) {
    cout << "\n" << DIVIDER << "\n";
    cout << "  SELECTION SORT\n";
    cout << DIVIDER << "\n";

    int n = arr.size();
    int iteration = 0;

    printArray(arr, "Data Awal :");
    cout << THIN << "\n";

    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            iteration++;
            if (arr[j] < arr[minIdx]) minIdx = j;
            cout << "  Pass " << i+1 << ", Step " << setw(2) << j-i
                 << " (iterasi " << setw(3) << iteration << "): "
                 << "bandingkan indeks " << i << "(" << arr[i] << ") dengan "
                 << j << "(" << arr[j] << ")  -> min sementara = a["
                 << minIdx << "]=" << arr[minIdx] << "\n";
        }
        if (minIdx != i) {
            cout << "  Tukar a[" << i << "]=" << arr[minIdx]
                 << " <-> a[" << minIdx << "]=" << arr[i] << "\n";
            swap(arr[i], arr[minIdx]);
        } else {
            cout << "  Tidak ada pertukaran (a[" << i << "]="
                 << arr[i] << " sudah minimum)\n";
        }
        cout << "  Setelah Pass " << i+1 << "    : ";
        printArrayRange(arr, 0, n-1);
        cout << "\n\n";
    }

    cout << THIN << "\n";
    cout << "  Total iterasi : " << iteration << "\n";
    printArray(arr, "Hasil Sorting :");
    return arr;
}

 /* ─── 3. INSERTION SORT ─── */

vector<int> insertionSort(vector<int> arr) {
    cout << "\n" << DIVIDER << "\n";
    cout << "  INSERTION SORT\n";
    cout << DIVIDER << "\n";

    int n = arr.size();
    int iteration = 0;

    printArray(arr, "Data Awal :");
    cout << THIN << "\n";

    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        cout << "  Pass " << i << ": ambil kunci = " << key << "\n";
        while (j >= 0 && arr[j] > key) {
            iteration++;
            arr[j + 1] = arr[j];
            cout << "    iterasi " << setw(3) << iteration
                 << ": geser a[" << j << "]=" << arr[j+1] << " ke kanan -> ";
            printArrayRange(arr, 0, n-1);
            cout << "\n";
            j--;
        }
        iteration++;
        arr[j + 1] = key;
        cout << "    iterasi " << setw(3) << iteration
             << ": sisipkan " << key << " di posisi " << j+1 << "  -> ";
        printArrayRange(arr, 0, n-1);
        cout << "\n\n";
    }

    cout << THIN << "\n";
    cout << "  Total iterasi : " << iteration << "\n";
    printArray(arr, "Hasil Sorting :");
    return arr;
}

 /* ─── 4. QUICK SORT ─── */

int qs_iter = 0;

int partitionQS(vector<int>& arr, int low, int high, int depth) {
    int pivot = arr[high];
    int i = low - 1;
    string indent(depth * 2, ' ');
    cout << "  " << indent << "Pivot = " << pivot << "  subarray = ";
    printArrayRange(arr, low, high);
    cout << "\n";

    for (int j = low; j < high; j++) {
        qs_iter++;
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
            cout << "  " << indent << "  iterasi " << setw(3) << qs_iter
                 << ": a[" << i << "] <-> a[" << j << "]  -> ";
            printArrayRange(arr, low, high);
            cout << "\n";
        } else {
            cout << "  " << indent << "  iterasi " << setw(3) << qs_iter
                 << ": lewati a[" << j << "]=" << arr[j]
                 << " (> pivot " << pivot << ")\n";
        }
        qs_iter++;
    }
    swap(arr[i + 1], arr[high]);
    cout << "  " << indent << "  Letakkan pivot " << pivot
         << " di posisi " << i+1 << "  -> ";
    printArrayRange(arr, low, high);
    cout << "\n";
    return i + 1;
}

void quickSortHelper(vector<int>& arr, int low, int high, int depth = 0) {
    if (low < high) {
        string indent(depth * 2, ' ');
        int pi = partitionQS(arr, low, high, depth);
        cout << "  " << indent << "Rekursif kiri  [" << low << ".." << pi-1 << "]: ";
        printArrayRange(arr, low, pi-1);
        cout << "\n";
        quickSortHelper(arr, low, pi - 1, depth + 1);
        cout << "  " << indent << "Rekursif kanan [" << pi+1 << ".." << high << "]: ";
        printArrayRange(arr, pi+1, high);
        cout << "\n";
        quickSortHelper(arr, pi + 1, high, depth + 1);
    }
}

vector<int> quickSort(vector<int> arr) {
    cout << "\n" << DIVIDER << "\n";
    cout << "  QUICK SORT\n";
    cout << DIVIDER << "\n";
    qs_iter = 0;

    printArray(arr, "Data Awal :");
    cout << THIN << "\n";

    quickSortHelper(arr, 0, arr.size() - 1);

    cout << THIN << "\n";
    cout << "  Total iterasi : " << qs_iter << "\n";
    printArray(arr, "Hasil Sorting :");
    return arr;
}

/* ─── 5. MERGE SORT ─── */

int ms_iter = 0;

void mergeArrays(vector<int>& arr, int left, int mid, int right, int depth) {
    string indent(depth * 2, ' ');
    vector<int> L(arr.begin() + left, arr.begin() + mid + 1);
    vector<int> R(arr.begin() + mid + 1, arr.begin() + right + 1);

    cout << "  " << indent << "Gabung: [ ";
    for (size_t i = 0; i < L.size(); i++) { cout << L[i]; if (i < L.size()-1) cout << ", "; }
    cout << " ] + [ ";
    for (size_t i = 0; i < R.size(); i++) { cout << R[i]; if (i < R.size()-1) cout << ", "; }
    cout << " ]\n";

    int i = 0, j = 0, k = left;
    while (i < (int)L.size() && j < (int)R.size()) {
        ms_iter++;
        if (L[i] <= R[j]) { arr[k++] = L[i++]; }
        else               { arr[k++] = R[j++]; }
    }
    while (i < (int)L.size()) arr[k++] = L[i++];
    while (j < (int)R.size()) arr[k++] = R[j++];

    cout << "  " << indent << "Hasil : ";
    printArrayRange(arr, left, right);
    cout << "\n";
}

void mergeSortHelper(vector<int>& arr, int left, int right, int depth = 0) {
    if (left < right) {
        string indent(depth * 2, ' ');
        int mid = (left + right) / 2;
        cout << "  " << indent << "Bagi [" << left << ".." << right
             << "] -> [" << left << ".." << mid
             << "] | [" << mid+1 << ".." << right << "]\n";
        mergeSortHelper(arr, left, mid, depth + 1);
        mergeSortHelper(arr, mid + 1, right, depth + 1);
        mergeArrays(arr, left, mid, right, depth);
    }
}

vector<int> mergeSort(vector<int> arr) {
    cout << "\n" << DIVIDER << "\n";
    cout << "  MERGE SORT\n";
    cout << DIVIDER << "\n";
    ms_iter = 0;

    printArray(arr, "Data Awal :");
    cout << THIN << "\n";

    mergeSortHelper(arr, 0, arr.size() - 1);

    cout << THIN << "\n";
    cout << "  Total iterasi : " << ms_iter << "\n";
    printArray(arr, "Hasil Sorting :");
    return arr;
}

/* ─── MAIN ─── */

int main() {
    cout << DIVIDER << "\n";
    cout << "  PROGRAM SORTING ALGORITHM - C++\n";
    cout << DIVIDER << "\n";

    int n;
    cout << "\n  Masukkan jumlah data : ";
    cin >> n;

    vector<int> data(n);
    cout << "  Masukkan " << n << " data : ";
    for (int i = 0; i < n; i++) cin >> data[i];

    cout << "\n  Data yang akan di-sort : ";
    printArrayRange(data, 0, n-1);
    cout << "\n";

    cout << "\n" << DIVIDER << "\n";
    cout << "  Pilih Algoritma Sorting:\n";
    cout << "  1. Bubble Sort\n";
    cout << "  2. Selection Sort\n";
    cout << "  3. Insertion Sort\n";
    cout << "  4. Quick Sort\n";
    cout << "  5. Merge Sort\n";
    cout << "  6. Semua Algoritma\n";
    cout << DIVIDER << "\n";
    cout << "  Masukkan pilihan (1-6) : ";

    int pilihan;
    cin >> pilihan;

    switch (pilihan) {
        case 1: bubbleSort(data);    break;
        case 2: selectionSort(data); break;
        case 3: insertionSort(data); break;
        case 4: quickSort(data);     break;
        case 5: mergeSort(data);     break;
        case 6:
            bubbleSort(data);
            selectionSort(data);
            insertionSort(data);
            quickSort(data);
            mergeSort(data);
            break;
        default:
            cout << "  Pilihan tidak valid.\n";
    }

    cout << "\n" << DIVIDER << "\n";
    cout << "  Program selesai.\n";
    cout << DIVIDER << "\n";

    return 0;
}
